luigi/worker.py

```diff
@@ -172,14 +172,25 @@ def run(self):
     def _recursive_terminate(self):
         import psutil
 
-        parent = psutil.Process(self.pid)
-        for child in parent.children(recursive=True):
-            child.kill()
+        try:
+            parent = psutil.Process(self.pid)
+            children = parent.children(recursive=True)
+
+            # terminate parent. Give it a chance to clean up
+            super(TaskProcess, self).terminate()
+            parent.wait()
```

Comment:
I think my idea here is that we're giving the parent the opportunity to
clean up after itself, before taking the more blunt approach of terminating
children. I can't remember if I hit an actual issue without this step, or
if it was just a precaution. I can rerun my problematic pipeline without
the wait, to check for any bad behavior.


On Thursday, September 10, 2015, Arash Rouhani <notifications@github.com>
wrote:

> In luigi/worker.py
> <https://github.com/spotify/luigi/pull/1191#discussion_r39225467>:
>
> > @@ -172,14 +172,25 @@ def run(self):
> >      def _recursive_terminate(self):
> >          import psutil
> >
> > -        parent = psutil.Process(self.pid)
> > -        for child in parent.children(recursive=True):
> > -            child.kill()
> > +        try:
> > +            parent = psutil.Process(self.pid)
> > +            children = parent.children(recursive=True)
> > +
> > +            # terminate parent. Give it a chance to clean up
> > +            super(TaskProcess, self).terminate()
> > +            parent.wait()
>
> Wait, is it really a good idea to block here? What if it's a process that
> never terminates even though we've "asked for it"? Won't the whole
> worker-loop block on this then? @ChrisBeaumont
> <https://github.com/ChrisBeaumont>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/luigi/pull/1191/files#r39225467>.
>